### PR TITLE
public API overhaul

### PR DESCRIPTION
**Pull request by OmarTawfik**
- rename `with_colour` to `with_color` as we are using a single `en-US` locale in the entire project.
- use `SUPPORTED_VERSIONS.binary_search()` in `Language::new()`
- use functions instead of properties for expensive operations -> #628
- require specifying an initial offset when creating a CST cursor -> #628

### CODE REVIEW

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs [hunk 1]
```diff
@@ -85,11 +85,14 @@ impl TestNodeBuilder {
 }
 
 impl TestNode {
-    pub fn from_cst(node: &Node) -> Self {
+    pub fn from_cst(cursor: Cursor) -> Self {
```
Comment (Xanewok) [inline crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:88]: You could use `mut cursor` here

File: crates/codegen/parser/runtime/src/support/parser_result.rs [hunk 2]
```diff
@@ -93,7 +93,7 @@ impl Match {
     pub fn is_full_recursive(&self) -> bool {
         self.nodes
             .iter()
-            .flat_map(cst::Node::cursor)
+            .flat_map(|node| cst::Node::create_cursor(node, Default::default()))
```
Comment (Xanewok) [inline crates/codegen/parser/runtime/src/support/parser_result.rs:96]: I'm not a fan of using the Default as parameter and think it'd be clearer if we have a separate `cursor_without_offset` (without the argument) function `cursor_with_offset` (or something along the lines).

This is fine for now, but I'd rethink how we can design the API to better convey that only the root node cursor has absolute offsets and other created ones are relative.

File: crates/codegen/parser/runtime/src/parse_output.rs [hunk 3]
```diff
@@ -18,4 +18,8 @@ impl ParseOutput {
     pub fn is_valid(&self) -> bool {
         return self.errors.is_empty();
     }
+
+    pub fn create_tree_cursor(&self) -> Cursor {
```
Comment (Xanewok) [inline crates/codegen/parser/runtime/src/parse_output.rs:22]: If exposing the API is important for us now, I would add more documentation to the public cursor functions that the user might use

**Review by Xanewok — COMMENTED** [review]
Left some comments, thanks for tackling the cursor offset API!

File: crates/codegen/parser/runtime/src/parse_output.rs (hunk 3 above)
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/parse_output.rs:22]: Sure!

File: crates/codegen/parser/runtime/src/support/parser_result.rs (hunk 2 above)
Comment (OmarTawfik) [reply crates/codegen/parser/runtime/src/support/parser_result.rs:96]: I think `cursor_without_offset()` should create a cursor that doesn't track offsets at all (for perf, and also to not expose an invalid `.offset` property).
Also I thought about introducing a `TextIndex::ZERO` helper.
But I think we should consider this as part of future updates to the API. I want to keep this PR minimal to address existing feedback.

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs (hunk 1 above)
Comment (OmarTawfik) [reply crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:88]: Done!

**Review by Xanewok — APPROVED** [review]

**Pull request merged**